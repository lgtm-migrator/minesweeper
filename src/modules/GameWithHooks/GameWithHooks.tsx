import { GameArea, WrapperContainer } from "@components/game-area/game-layout";
import { GameOver } from "@components/game-area/game-over";
import { GridComponent } from "@components/grid";
import { ScoreBoard } from "@components/scoreboard";
import { Top } from "@components/top-section";
import { GameLevels, LevelNames } from "@modules/GameSettings";
import { useGame } from "@modules/GameWithHooks/hooks/use-game";
import React, { FunctionComponent } from "react";

export const GameWithHooks: FunctionComponent = () => {
  const {
    level,
    isWin,
    isGameOver,
    settings,
    playerField,
    onContextMenuHandler,
    onChangeLevelHandler,
    onClickHandler,
    onResetHandler,
  } = useGame();

  const [, bombs] = settings;

  return (
    <WrapperContainer>
      <Top feature="Flag" firstAction="right click">
        Minesweeper
      </Top>
      <GameArea>
        <ScoreBoard
          time="000"
          mines={String(bombs)}
          defaultLevel={level}
          levels={GameLevels as unknown as string[]}
          onReset={onResetHandler}
          onChangeLevel={({ target: { value: level } }: React.ChangeEvent<HTMLSelectElement>) =>
            onChangeLevelHandler(level as LevelNames)
          }
        />
        {isGameOver && <GameOver onClick={onResetHandler} isWin={isWin} />}
        <GridComponent children={playerField} onClick={onClickHandler} onContextMenu={onContextMenuHandler} />
      </GameArea>
    </WrapperContainer>
  );
};
