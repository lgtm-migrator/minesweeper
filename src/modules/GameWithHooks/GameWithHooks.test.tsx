import { GameWithHooks } from "@modules/GameWithHooks/GameWithHooks";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";

describe("GameWithHooks test cases", function () {
  describe("Render component behaviour", function () {
    it("should render game field by default", function () {
      const { asFragment } = render(<GameWithHooks />);

      expect(screen.getAllByRole("cell")).toHaveLength(81);

      expect(asFragment()).toMatchSnapshot();
    });

    it("should observe onChange handler on game level", async function () {
      render(<GameWithHooks />);

      expect(screen.getAllByRole("cell")).toHaveLength(81);
      await userEvent.selectOptions(screen.getByRole("select-component"), "intermediate");
      expect(screen.getByRole("option", { name: "intermediate" })).toBeEnabled();
      expect(screen.getAllByRole("cell")).toHaveLength(256);

      await userEvent.selectOptions(screen.getByRole("select-component"), "expert");
      expect(screen.getAllByRole("cell")).toHaveLength(484);
    });
  });
});
