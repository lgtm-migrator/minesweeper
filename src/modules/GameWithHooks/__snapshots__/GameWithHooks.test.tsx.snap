// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameWithHooks test cases Render component behaviour should render game field by default 1`] = `
<DocumentFragment>
  <div
    class="wrapperContainer"
  >
    <header
      class="headerContainer"
    >
      <h1
        class="gameName"
      >
        Minesweeper
      </h1>
      <legend
        class="legendParent"
      >
        <strong>
          Flag: 
        </strong>
        <code
          class="flagComboParent"
        >
          <span
            class="firstAction"
          >
            right click
          </span>
        </code>
      </legend>
    </header>
    <div
      class="gameArea"
    >
      <div
        class="wrapperScoreboard"
      >
        <div
          class="parentCounter"
        >
          000
        </div>
        <div>
          <select
            class="selectLevel"
            role="select-component"
          >
            <option
              class="optionLevel"
            >
              beginner
            </option>
            <option
              class="optionLevel"
            >
              intermediate
            </option>
            <option
              class="optionLevel"
            >
              expert
            </option>
          </select>
          <button
            class="resetButton"
          >
            ðŸ™‚
          </button>
        </div>
        <div
          class="parentCounter"
        >
          10
        </div>
      </div>
      <div
        class="containerWrapper"
        role="grid"
        style="grid-template-columns: repeat(9, auto);"
      >
        <div
          aria-label="0,0"
          class="closedFrame changedBorderColor"
          data-testid="0,0"
          role="cell"
        />
        <div
          aria-label="0,1"
          class="closedFrame changedBorderColor"
          data-testid="0,1"
          role="cell"
        />
        <div
          aria-label="0,2"
          class="closedFrame changedBorderColor"
          data-testid="0,2"
          role="cell"
        />
        <div
          aria-label="0,3"
          class="closedFrame changedBorderColor"
          data-testid="0,3"
          role="cell"
        />
        <div
          aria-label="0,4"
          class="closedFrame changedBorderColor"
          data-testid="0,4"
          role="cell"
        />
        <div
          aria-label="0,5"
          class="closedFrame changedBorderColor"
          data-testid="0,5"
          role="cell"
        />
        <div
          aria-label="0,6"
          class="closedFrame changedBorderColor"
          data-testid="0,6"
          role="cell"
        />
        <div
          aria-label="0,7"
          class="closedFrame changedBorderColor"
          data-testid="0,7"
          role="cell"
        />
        <div
          aria-label="0,8"
          class="closedFrame changedBorderColor"
          data-testid="0,8"
          role="cell"
        />
        <div
          aria-label="1,0"
          class="closedFrame changedBorderColor"
          data-testid="1,0"
          role="cell"
        />
        <div
          aria-label="1,1"
          class="closedFrame changedBorderColor"
          data-testid="1,1"
          role="cell"
        />
        <div
          aria-label="1,2"
          class="closedFrame changedBorderColor"
          data-testid="1,2"
          role="cell"
        />
        <div
          aria-label="1,3"
          class="closedFrame changedBorderColor"
          data-testid="1,3"
          role="cell"
        />
        <div
          aria-label="1,4"
          class="closedFrame changedBorderColor"
          data-testid="1,4"
          role="cell"
        />
        <div
          aria-label="1,5"
          class="closedFrame changedBorderColor"
          data-testid="1,5"
          role="cell"
        />
        <div
          aria-label="1,6"
          class="closedFrame changedBorderColor"
          data-testid="1,6"
          role="cell"
        />
        <div
          aria-label="1,7"
          class="closedFrame changedBorderColor"
          data-testid="1,7"
          role="cell"
        />
        <div
          aria-label="1,8"
          class="closedFrame changedBorderColor"
          data-testid="1,8"
          role="cell"
        />
        <div
          aria-label="2,0"
          class="closedFrame changedBorderColor"
          data-testid="2,0"
          role="cell"
        />
        <div
          aria-label="2,1"
          class="closedFrame changedBorderColor"
          data-testid="2,1"
          role="cell"
        />
        <div
          aria-label="2,2"
          class="closedFrame changedBorderColor"
          data-testid="2,2"
          role="cell"
        />
        <div
          aria-label="2,3"
          class="closedFrame changedBorderColor"
          data-testid="2,3"
          role="cell"
        />
        <div
          aria-label="2,4"
          class="closedFrame changedBorderColor"
          data-testid="2,4"
          role="cell"
        />
        <div
          aria-label="2,5"
          class="closedFrame changedBorderColor"
          data-testid="2,5"
          role="cell"
        />
        <div
          aria-label="2,6"
          class="closedFrame changedBorderColor"
          data-testid="2,6"
          role="cell"
        />
        <div
          aria-label="2,7"
          class="closedFrame changedBorderColor"
          data-testid="2,7"
          role="cell"
        />
        <div
          aria-label="2,8"
          class="closedFrame changedBorderColor"
          data-testid="2,8"
          role="cell"
        />
        <div
          aria-label="3,0"
          class="closedFrame changedBorderColor"
          data-testid="3,0"
          role="cell"
        />
        <div
          aria-label="3,1"
          class="closedFrame changedBorderColor"
          data-testid="3,1"
          role="cell"
        />
        <div
          aria-label="3,2"
          class="closedFrame changedBorderColor"
          data-testid="3,2"
          role="cell"
        />
        <div
          aria-label="3,3"
          class="closedFrame changedBorderColor"
          data-testid="3,3"
          role="cell"
        />
        <div
          aria-label="3,4"
          class="closedFrame changedBorderColor"
          data-testid="3,4"
          role="cell"
        />
        <div
          aria-label="3,5"
          class="closedFrame changedBorderColor"
          data-testid="3,5"
          role="cell"
        />
        <div
          aria-label="3,6"
          class="closedFrame changedBorderColor"
          data-testid="3,6"
          role="cell"
        />
        <div
          aria-label="3,7"
          class="closedFrame changedBorderColor"
          data-testid="3,7"
          role="cell"
        />
        <div
          aria-label="3,8"
          class="closedFrame changedBorderColor"
          data-testid="3,8"
          role="cell"
        />
        <div
          aria-label="4,0"
          class="closedFrame changedBorderColor"
          data-testid="4,0"
          role="cell"
        />
        <div
          aria-label="4,1"
          class="closedFrame changedBorderColor"
          data-testid="4,1"
          role="cell"
        />
        <div
          aria-label="4,2"
          class="closedFrame changedBorderColor"
          data-testid="4,2"
          role="cell"
        />
        <div
          aria-label="4,3"
          class="closedFrame changedBorderColor"
          data-testid="4,3"
          role="cell"
        />
        <div
          aria-label="4,4"
          class="closedFrame changedBorderColor"
          data-testid="4,4"
          role="cell"
        />
        <div
          aria-label="4,5"
          class="closedFrame changedBorderColor"
          data-testid="4,5"
          role="cell"
        />
        <div
          aria-label="4,6"
          class="closedFrame changedBorderColor"
          data-testid="4,6"
          role="cell"
        />
        <div
          aria-label="4,7"
          class="closedFrame changedBorderColor"
          data-testid="4,7"
          role="cell"
        />
        <div
          aria-label="4,8"
          class="closedFrame changedBorderColor"
          data-testid="4,8"
          role="cell"
        />
        <div
          aria-label="5,0"
          class="closedFrame changedBorderColor"
          data-testid="5,0"
          role="cell"
        />
        <div
          aria-label="5,1"
          class="closedFrame changedBorderColor"
          data-testid="5,1"
          role="cell"
        />
        <div
          aria-label="5,2"
          class="closedFrame changedBorderColor"
          data-testid="5,2"
          role="cell"
        />
        <div
          aria-label="5,3"
          class="closedFrame changedBorderColor"
          data-testid="5,3"
          role="cell"
        />
        <div
          aria-label="5,4"
          class="closedFrame changedBorderColor"
          data-testid="5,4"
          role="cell"
        />
        <div
          aria-label="5,5"
          class="closedFrame changedBorderColor"
          data-testid="5,5"
          role="cell"
        />
        <div
          aria-label="5,6"
          class="closedFrame changedBorderColor"
          data-testid="5,6"
          role="cell"
        />
        <div
          aria-label="5,7"
          class="closedFrame changedBorderColor"
          data-testid="5,7"
          role="cell"
        />
        <div
          aria-label="5,8"
          class="closedFrame changedBorderColor"
          data-testid="5,8"
          role="cell"
        />
        <div
          aria-label="6,0"
          class="closedFrame changedBorderColor"
          data-testid="6,0"
          role="cell"
        />
        <div
          aria-label="6,1"
          class="closedFrame changedBorderColor"
          data-testid="6,1"
          role="cell"
        />
        <div
          aria-label="6,2"
          class="closedFrame changedBorderColor"
          data-testid="6,2"
          role="cell"
        />
        <div
          aria-label="6,3"
          class="closedFrame changedBorderColor"
          data-testid="6,3"
          role="cell"
        />
        <div
          aria-label="6,4"
          class="closedFrame changedBorderColor"
          data-testid="6,4"
          role="cell"
        />
        <div
          aria-label="6,5"
          class="closedFrame changedBorderColor"
          data-testid="6,5"
          role="cell"
        />
        <div
          aria-label="6,6"
          class="closedFrame changedBorderColor"
          data-testid="6,6"
          role="cell"
        />
        <div
          aria-label="6,7"
          class="closedFrame changedBorderColor"
          data-testid="6,7"
          role="cell"
        />
        <div
          aria-label="6,8"
          class="closedFrame changedBorderColor"
          data-testid="6,8"
          role="cell"
        />
        <div
          aria-label="7,0"
          class="closedFrame changedBorderColor"
          data-testid="7,0"
          role="cell"
        />
        <div
          aria-label="7,1"
          class="closedFrame changedBorderColor"
          data-testid="7,1"
          role="cell"
        />
        <div
          aria-label="7,2"
          class="closedFrame changedBorderColor"
          data-testid="7,2"
          role="cell"
        />
        <div
          aria-label="7,3"
          class="closedFrame changedBorderColor"
          data-testid="7,3"
          role="cell"
        />
        <div
          aria-label="7,4"
          class="closedFrame changedBorderColor"
          data-testid="7,4"
          role="cell"
        />
        <div
          aria-label="7,5"
          class="closedFrame changedBorderColor"
          data-testid="7,5"
          role="cell"
        />
        <div
          aria-label="7,6"
          class="closedFrame changedBorderColor"
          data-testid="7,6"
          role="cell"
        />
        <div
          aria-label="7,7"
          class="closedFrame changedBorderColor"
          data-testid="7,7"
          role="cell"
        />
        <div
          aria-label="7,8"
          class="closedFrame changedBorderColor"
          data-testid="7,8"
          role="cell"
        />
        <div
          aria-label="8,0"
          class="closedFrame changedBorderColor"
          data-testid="8,0"
          role="cell"
        />
        <div
          aria-label="8,1"
          class="closedFrame changedBorderColor"
          data-testid="8,1"
          role="cell"
        />
        <div
          aria-label="8,2"
          class="closedFrame changedBorderColor"
          data-testid="8,2"
          role="cell"
        />
        <div
          aria-label="8,3"
          class="closedFrame changedBorderColor"
          data-testid="8,3"
          role="cell"
        />
        <div
          aria-label="8,4"
          class="closedFrame changedBorderColor"
          data-testid="8,4"
          role="cell"
        />
        <div
          aria-label="8,5"
          class="closedFrame changedBorderColor"
          data-testid="8,5"
          role="cell"
        />
        <div
          aria-label="8,6"
          class="closedFrame changedBorderColor"
          data-testid="8,6"
          role="cell"
        />
        <div
          aria-label="8,7"
          class="closedFrame changedBorderColor"
          data-testid="8,7"
          role="cell"
        />
        <div
          aria-label="8,8"
          class="closedFrame changedBorderColor"
          data-testid="8,8"
          role="cell"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
